File: halalit/www/halalit-lookup-quality.js
/**
 * Halalit — detect garbled scroll-scanner / OCR title+author reads.
 */
(function (global) {
  var PAREN_FRAGMENT_RE = /(?:^\s*[A-Za-z]\)|\(\s*[A-Za-z]\s*\)|\)\s*[a-z]{1,3}\s*$)/;

  function coreWord(w) {
    return String(w || "").replace(/[^A-Za-z'.-]/g, "");
  }

  function lineGarbage(line, kind) {
    line = String(line || "").trim();
    if (!line) return false;
    if (PAREN_FRAGMENT_RE.test(line)) return true;
    var letters = line.replace(/[^A-Za-z]/g, "");
    if (letters.length < 3) return true;
    if (letters.length / Math.max(line.length, 1) < 0.55) return true;
    var words = line.split(/\s+/).filter(Boolean);
    if (!words.length) return true;

    var realWords = 0;
    var longWords = 0;
    for (var i = 0; i < words.length; i++) {
      var core = coreWord(words[i]);
      if (core.length >= 2 && /[aeiouAEIOU]/.test(core)) realWords++;
      if (core.length >= 4 && /[aeiouAEIOU]/.test(core)) longWords++;
    }
    if (!realWords) return true;

    var cores = [];
    for (var j = 0; j < words.length; j++) {
      var c = coreWord(words[j]);
      if (c) cores.push(c);
    }
    if (cores.length && cores.every(function (c) {
      return c.length <= 3;
    })) {
      return true;
    }

    if (kind === "title") {
      if (longWords === 0 && words.join(" ").length < 14) return true;
    }
    if (kind === "author") {
      if (longWords === 0 && words.length <= 2 && letters.length < 8) return true;
    }
    return false;
  }

  function isGarbage(title, author) {
    title = String(title || "").trim();
    author = String(author || "").trim();
    if (!title) return true;
    if (lineGarbage(title, "title")) return true;
    if (author && lineGarbage(author, "author")) return true;
    return false;
  }

  function reportMalfunction(title, author, phase) {
    var base =
      global.HalalitAccountStorage && global.HalalitAccountStorage.apiBase
        ? global.HalalitAccountStorage.apiBase()
        : global.HalalitBookcheckConfig && global.HalalitBookcheckConfig.apiBase
          ? global.HalalitBookcheckConfig.apiBase()
          : "";
    if (!base || !global.fetch) return;
    var key = "halalitScannerAlert:" + phase + ":" + title + "|" + author;
    try {
      if (global.sessionStorage && global.sessionStorage.getItem(key)) return;
      if (global.sessionStorage) global.sessionStorage.setItem(key, "1");
    } catch (e) {}
    global.fetch(base + "/scanner/malfunction-report", {
      method: "POST",
      credentials: "include",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        title: title,
        author: author || "",
        phase: phase || "cover_read",
      }),
    }).catch(function () {});
  }

  global.HalalitLookupQuality = {
    isGarbage: isGarbage,
    reportMalfunction: reportMalfunction,
  };
})(typeof window !== "undefined" ? window : this);

File: halalit/server/halalit_lookup_quality.py
"""Filter garbled cover-OCR / scroll-scanner title+author pairs from owner lookup lists."""
from __future__ import annotations

import re

_WORD_RE = re.compile(r"\s+")
_PAREN_FRAGMENT_RE = re.compile(r"(?:^\s*[A-Za-z]\)|\(\s*[A-Za-z]\s*\)|\)\s*[a-z]{1,3}\s*$)")


def _words(line: str) -> list[str]:
    return [w for w in _WORD_RE.split(line.strip()) if w]


def _core_word(word: str) -> str:
    return re.sub(r"[^A-Za-z'.-]", "", word)


def _line_garbage(line: str, kind: str) -> bool:
    line = (line or "").strip()
    if not line:
        return False
    if _PAREN_FRAGMENT_RE.search(line):
        return True
    letters = re.sub(r"[^A-Za-z]", "", line)
    if len(letters) < 3:
        return True
    if len(letters) / max(len(line), 1) < 0.55:
        return True
    words = _words(line)
    if not words:
        return True

    real_words = 0
    long_words = 0
    for w in words:
        core = _core_word(w)
        if len(core) >= 2 and re.search(r"[aeiouAEIOU]", core):
            real_words += 1
        if len(core) >= 4 and re.search(r"[aeiouAEIOU]", core):
            long_words += 1
    if not real_words:
        return True

    cores = [_core_word(w) for w in words if _core_word(w)]
    if cores and all(len(c) <= 3 for c in cores):
        return True

    if kind == "title":
        if long_words == 0 and len(" ".join(words)) < 14:
            return True
    if kind == "author":
        if long_words == 0 and len(words) <= 2 and len(letters) < 8:
            return True

    return False


def is_garbage_lookup(title: str, author: str = "") -> bool:
    title = (title or "").strip()
    author = (author or "").strip()
    if not title:
        return True
    if _line_garbage(title, "title"):
        return True
    if author and _line_garbage(author, "author"):
        return True
    return False
